pkg/discovery/mesos_leader.go

```diff
@@ -0,0 +1,243 @@
+func (mesosLeader *MesosLeader) RefreshMesosLeaderLogin() (error) {
+	glog.V(3).Infof("RefreshMesosLeaderLogin %++v", mesosLeader.leaderConf)
+	// API request to Login to the current Mesos Master leader and save the login token for subsequent discovery requests
+	token, err := mesosLeader.leaderRestClient.Login()
+	if err == nil {
+		glog.V(3).Infof("Mesos login api succeeded with existing leader : %++v\n", mesosLeader.leaderConf)
+		mesosLeader.leaderConf.Token = token
+		return nil
+	}
+
+	glog.Errorf("Error logging to Mesos Leader at %s, detecting new leader using : %s",
+			mesosLeader.leaderConf.MasterIP, mesosLeader.targetConf.MasterIPPort)
+	// Refresh the leader
+	err = mesosLeader.updateMesosLeader()
+	return err
+}
+
+// Get the RestAPI Client using the given master config.
+// Return MasterRestClient if login is successful, nil if the login fails
+func (mesosLeader *MesosLeader) getRestAPIClient(targetConf *conf.MesosTargetConf, masterConf *conf.MasterConf) (master.MasterRestClient,  error) {
+	// Bad master config
+	if masterConf.MasterIP == "" {
+		return nil, fmt.Errorf("Null master ip : " + string(targetConf.Master))
+	}
+	var masterRestClient master.MasterRestClient
+
+	// Find if the api client exists for this master
+	mapkey := strings.Join([]string{masterConf.MasterIP, masterConf.MasterPort}, ":")
+
+	masterRestClient, exists := mesosLeader.restApiClientMap[MASTER_IP_PORT(mapkey)]
+	if !exists || masterRestClient  == nil {	// does not exist, create new
+		// supply credentials from the target conf before creating the login request
+		masterConf.MasterUsername = targetConf.MasterUsername
+		masterConf.MasterPassword = targetConf.MasterPassword
+		masterConf.Master = targetConf.Master
+		glog.V(3).Infof("Creating new master rest api client %++v", masterConf)
+		// Create a new rest api client using the given master conf
+		masterRestClient = master.GetMasterRestClient(targetConf.Master, masterConf)
+		mesosLeader.restApiClientMap[MASTER_IP_PORT(mapkey)] = masterRestClient
+	}
+
+	if masterRestClient == nil {	//does not exist and cannot create new instance
+		glog.Errorf("Cannot find rest api client for master %s:%s::%s",
+					string(targetConf.Master))
+		return nil, fmt.Errorf("Cannot find rest api client for master %s:%s::%s ",
+					string(targetConf.Master))
+	}
+
+	// Login to the Mesos Master and save the login token
+	token, err := masterRestClient.Login()
+	if err != nil {
+		glog.Errorf("Error logging to mesos master at %s::%s : %s " +
+				masterConf.MasterIP, masterConf.MasterPort, err)
+		return nil, fmt.Errorf("Error logging to mesos master at %s::%s : %s " +
+					masterConf.MasterIP, masterConf.MasterPort, err)
+	}
+	// Save the token
+	masterConf.Token = token
+	return masterRestClient, nil
+}
+
+func parseMasterIPPorts(ipportlist string) []*conf.MasterConf {
+	// Split on comma.
+	ipports := strings.Split(ipportlist, ",")
+
+	var masterConfList []*conf.MasterConf
+	for i := range ipports {
```

Comment:
seems more concise in:

for _, ipport := range ipports {
    result := strings.Split(ipport, ":")
}